packages/cli/ts/commands/checkVerifyingKeys.ts

```diff
@@ -68,38 +64,43 @@ export const checkVerifyingKeys = async ({
   }
 
   // extract the verification keys from the zkey files
-  const processVk = VerifyingKey.fromObj(await extractVk(processMessagesZkeyPath));
-  const tallyVk = VerifyingKey.fromObj(await extractVk(tallyVotesZkeyPath));
-  const pollJoiningVk = VerifyingKey.fromObj(await extractVk(pollJoiningZkeyPath));
-  const pollJoinedVk = VerifyingKey.fromObj(await extractVk(pollJoinedZkeyPath));
+  const { pollJoiningVk, pollJoinedVk, processVk, tallyVk } = await extractAllVks({
+    processMessagesZkeyPath,
+    tallyVotesZkeyPath,
+    pollJoiningZkeyPath,
+    pollJoinedZkeyPath,
+  });
 
   try {
     logYellow(quiet, info("Retrieving verifying keys from the contract..."));
     // retrieve the verifying keys from the contract
 
     const mode = useQuadraticVoting ? EMode.QV : EMode.NON_QV;
 
-    const [pollJoiningVkOnChain, pollJoinedVkOnChain, processVkOnChain, tallyVkOnChain] = await Promise.all([
-      vkRegistryContractInstance.getPollJoiningVk(stateTreeDepth, voteOptionTreeDepth),
-      vkRegistryContractInstance.getPollJoinedVk(stateTreeDepth, voteOptionTreeDepth),
-      vkRegistryContractInstance.getProcessVk(stateTreeDepth, voteOptionTreeDepth, messageBatchSize, mode),
-      vkRegistryContractInstance.getTallyVk(stateTreeDepth, intStateTreeDepth, voteOptionTreeDepth, mode),
-    ]);
+    const { pollJoiningVkOnChain, pollJoinedVkOnChain, processVkOnChain, tallyVkOnChain } = await getAllOnChainVks({
+      vkRegistryAddress: vkContractAddress,
+      signer,
+      stateTreeDepth,
+      voteOptionTreeDepth,
+      messageBatchSize,
+      intStateTreeDepth,
+      mode,
+    });
 
     // do the actual validation
-    if (!compareVks(pollJoiningVk, pollJoiningVkOnChain)) {
+    if (!compareVks(pollJoiningVk!, pollJoiningVkOnChain)) {
```

Comment:
I think it makes sense to allow passing undefined value and check inside `compareVks` if value is defined.

Reply:
but if it's not defined then what do we do? return true?

Reply:
maybe makes sense to actually check first in checkVerifyingKeys if (pollJoiningvk) { compareVk(poll...)}?

Reply:
either way the caller will need to check after if the result is false and not handle it if the vk is undefined, so why do that? 

I think it's better to keep as is, then when calling do:

if (vk) { compare(vk, vkOnChain) }

otherwise it'd need to be

const res = compare(vk, vkOnChain)
if (!res && !vk) { do nothing }

or am I missing something?

Reply:
If you pass undefined to compare it will fail and go to catch with ic is not exist in undefined.
And there is no case where you don't have key but it exists onchain.
So my idea is to update compare function so it looks like this:

```
export const compare = (vk?: Vk, onchainVk: Vk): boolean => {
  if (!vk) {
    throw new Error("No vk provided")
  }
}
```

This will show more readable error and the logic is the same.